---
title: "BCB420 Assignment 1"
subtitle: "Dataset Selection and Initial Processing" 
author: "Maryam Hasanzadehkiabi"
date: "r Sys.Date()"
output:
    html_document:
        toc: TRUE
        toc_depth: 2
bibliography: assignment_1_draft.bib
csl: american-medical-association.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Dataset selection

## Insert required packages

```{r message=FALSE, warning=FALSE}
if (!requireNamespace("rmarkdown", quietly = TRUE))
    +     install.packages("rmarkdown")
if (!requireNamespace("knitr", quietly = TRUE))
    +     install.packages("knitr")
if (!requireNamespace("pandoc", quietly = TRUE))
    +     install.packages("pandoc")
if (!requireNamespace("devtools", quietly = TRUE))
    +     install.packages("devtools")
if (!requireNamespace("BiocManager", quietly = TRUE))
    +     install.packages("BiocManager")
if (!requireNamespace("edgeR", quietly = TRUE))
    BiocManager::install("edgeR")
if (!requireNamespace("biomaRt", quietly = TRUE))
    BiocManager::install("biomaRt")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")
if (!requireNamespace("GEOquery", quietly = TRUE))
    +     install.packages("GEOquery")
if (!requireNamespace("DBI", quietly = TRUE))
    +     install.packages("DBI")
if (!requireNamespace("RSQLite", quietly = TRUE))
    +     install.packages("RSQLite")
if(!requireNamespace("kableExtra", quietly=TRUE))
    install.packages("kableExtra")
```

## Run libraries

```{r message=FALSE, warning=FALSE}
library(devtools)
library(rmarkdown)
library(knitr)
library(pandoc)
library(BiocManager)
library(GEOmetadb)
library(GEOquery)
library(DBI)
library(RSQLite)
library(edgeR)
library(biomaRt)
library(kableExtra)
```

## Selection criteria

*RNASeq data
*Within the last 10 years
*Limited to human
*Number of samples more than 6
*Related to Metabolic Syndrome

```{r message=FALSE, warning=FALSE}
sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
             " gse.submission_date,",
             " gse.supplementary_file", 
             "FROM",
             "  gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
             "  JOIN gpl ON gse_gpl.gpl=gpl.gpl",
             "WHERE",
             "  gse.submission_date > '2013-01-01' AND", 
             "  gse.title LIKE '%Metabolic Syndrome%' AND", 
             "  gpl.organism LIKE '%Homo sapiens%' AND",
             "  gpl.technology LIKE '%high-throughput sequencing%' ",
             "  ORDER BY gse.submission_date DESC",sep=" ")
```

## Selecting samples with counts data

```{r}
if(!file.exists('GEOmetadb.sqlite')) getSQLiteFile()
con <- dbConnect(SQLite(),'GEOmetadb.sqlite')
rs <- dbGetQuery(con,sql)
counts_files <- rs$supplementary_file[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

series_of_interest <- rs$gse[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

shortened_filenames <- unlist(lapply(counts_files,
             FUN = function(x){x <- unlist(strsplit(x,";")) ;
              x <- x[grep(x,pattern= "count|cnt",ignore.case = TRUE)];
               tail(unlist(strsplit(x,"/")),n=1)}))
shortened_filenames[1:10]
```
The output of the interested series resulted in only one series that holds the selection criteria. Therefore, in the following steps, GSE166474 is being checked to find whether it contains sufficient samples to continue further analyses.

## Details of the GSE166474 series

Number of entities
```{r}
gse <- getGEO("GSE166474",GSEMatrix=FALSE)
```
Data owner contact information
```{r}
kable(data.frame(head(Meta(gse))), format = "pipe")
```
Name and number of the samples
```{r}
names(GSMList(gse))
```
More information about the series
```{r}
show(getGEO("GSE166474",GSEMatrix=TRUE))
```
Samples classification
```{r}
data.frame(getGEO("GSE166474",GSEMatrix=TRUE)[[1]])[,c(8)]
```
## Platform details
```{r}
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))
```
**Platform accession number:**
```{r echo=FALSE}
current_gpl_info$geo_accession
```
**Platform title:**
```{r echo=FALSE}
current_gpl_info$title
```
**Submission date:**
```{r echo=FALSE}
current_gpl_info$submission_date
```
**Last update date:**
```{r echo=FALSE}
current_gpl_info$last_update_date
```
**Organisms:**
```{r echo=FALSE}
current_gpl_info$organism
```
**Technology:**
```{r echo=FALSE}
current_gpl_info$technology
```
**Number of GEO datasets that use this technology:*
```{r echo=FALSE}
length(current_gpl_info$series_id)
```
**Number of GEO samples that use this technology:**
```{r echo=FALSE}
length(current_gpl_info$sample_id)
```

# 2. Data expression

## Supplementary files

```{r echo=FALSE}
supp_files = getGEOSuppFiles('GSE166474')
file_names = rownames(supp_files)
print(file_names)
```

Displaying data included in the first 10 rows collected over 3 metabolic syndrome patients. 

```{r}
genes_exp = read.delim(file_names[1],header=TRUE,
          check.names = FALSE)
genes_exp[1:10, 1:4] %>%
    kableExtra::kbl(caption = "Genes Expression in 3 samples") %>%
    kableExtra::kable_classic(full_width = F, html_font = "Cambria")
```

Number of the genes and name of the variables included in the dataset

```{r}
dim(genes_exp)
colnames(genes_exp)[1] <- "ensembl_id"
colnames(genes_exp)
```
## Samples category

Grouping samples. There are 2 groups, "M" indicates metabolic syndrome patients and "N" indicates healthy (normal) individuals. Each group contains 5 samples.

```{r}
samples <- data.frame(lapply(colnames(genes_exp)[2:11],
                             FUN=function(x){unlist(strsplit(x,
                            split = ""))[c(1)]}))
colnames(samples) <- colnames(genes_exp)[2:11]
rownames(samples) <- c("Status")
samples <- data.frame(t(samples))
```
## Genes filtering

Removing genes that do not follow ensembl nomenclature 
```{r}
valid_genes <- genes_exp[grep("ENSG", genes_exp$ensembl_id),]
```
Checking potential duplication. If the logical output returns any "TRUE" value, there does exist duplication. Otherwise the set contains unique genes. 
```{r}
summary(duplicated(valid_genes$ensembl_id))
```
Noise filtering to keep only genes with at least 1 read per million in 5 samples. 
```{r}
cpms = cpm(valid_genes[, 2:11])
rownames(cpms) <- valid_genes[,1]
keep = rowSums(cpms >1) >=5
genes_filtered = valid_genes[keep,]
filtered_results <- data.frame(genes_exp = nrow(genes_exp), valid_genes = nrow(valid_genes), genes_filtered = nrow(genes_filtered))
rownames(filtered_results)[1] <- "Number of Genes"
filtered_results %>%
  kableExtra::kbl(caption = "Genes count") %>%
  kableExtra::kable_classic(full_width = F, html_font = "Cambria")
```
## Displaying the distribution of data

Data distribution Boxplot
```{r warning=FALSE}
data2plot <- log2(cpm(genes_filtered[,2:11]))
colnames(data2plot) <- lapply(colnames(genes_filtered)[2:11],
                              FUN=function(x) {unlist(strsplit(x,
                              split="\\_"))[c(1)]})
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
las = 2, cex = 0.5, cex.lab = 0.5,
cex.axis = 0.5, main = "Data distribution - Boxplot")
abline(h = median(apply(data2plot, 2, median)),
col = "red", lwd = 1.2, lty = "dashed")
```
Data distribution Density plot
```{r}
counts_density <- apply(log2(cpm(genes_filtered[,2:11])),
2, density)
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
xlim <- range(c(xlim, counts_density[[i]]$x));
ylim <- range(c(ylim, counts_density[[i]]$y))
}
cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
ylab="Smoothing density of log2-CPM",
main="Data distribution - Density Plot", cex.lab = 0.85)
for (i in 1:length(counts_density))
lines(counts_density[[i]], col=cols[i], lty=ltys[i])
legend("topright", colnames(data2plot),
col=cols, lty=ltys, cex=0.75,
border ="blue", text.col = "black",
merge = TRUE, bg = "gray90")
```
MA plot for data comparing metabolic syndrome vs control groups

```{r}
plotMA(log2(genes_filtered[,c(2:6,7:11)]), ylab="M - ratio log expression",
main="MetS vs Control")
```
## Normalization

Trimmed Mean of M-values (TMM)

```{r}
filtered_data_matrix <- as.matrix(genes_filtered[,2:11])
rownames(filtered_data_matrix) <- genes_filtered$ensembl_id
d = DGEList(counts=filtered_data_matrix, group=samples$Status)
#d_DGE = DGEList(counts=filtered_data_matrix, group=samples$Status)
```
```{r}
d = calcNormFactors(d)
#d_NormFac = calcNormFactors(d_DGE)
```
```{r}
normalized_counts <- cpm(d)
#normalized_counts <- cpm(d_NormFac)
```
```{r}
plotMDS(d, labels=lapply(rownames(samples), FUN=function(x){unlist(strsplit(x,split = "\\_"))[c(1)]}), col = c("darkgreen","blue")[factor(samples$Status)])
#plotMDS(d_NormFac, labels=lapply(rownames(samples), #FUN=function(x){unlist(strsplit(x,split = "\\_"))[c(1)]}), col = #c("darkgreen","blue")[factor(samples$Status)])
```
```{r}
model_design <- model.matrix(~samples$Status)
```
```{r}
d <- estimateDisp(d, model_design)
#d_Disp <- estimateDisp(d_NormFac, model_design)
```
```{r}
plotBCV(d,col.tagwise = "black",col.common = "red",col.trend = "blue")
#plotBCV(d_Disp,col.tagwise = "black",col.common = "red",)
```
# Mapping

```{r}
ensembl <- useMart("ensembl")
datasets = listDatasets(ensembl)
kable(head(datasets),format = "pipe")
kable(head(datasets[grep(datasets$dataset,
                         pattern = "sapiens"),]),format = "pipe")
ensembl = useDataset("hsapiens_gene_ensembl",mart=ensembl)
dim(listFilters(ensembl))
#ensembl = useEnsembl(biomart='ensembl', dataset=ensembl_dataset, mirror = "uswest")
```
```{r}
biomart_human_filters <- listFilters(ensembl)
kable(biomart_human_filters[
grep(biomart_human_filters$name,pattern="ensembl"),],
format="html") %>%
row_spec(3, background = "yellow")
```

```{r}
gene_mapped <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
filters = c("ensembl_gene_id"),
values = genes_filtered$ensembl_id,
mart = ensembl)
```

```{r}
length(which(rownames(normalized_counts) %in%
gene_mapped$ensembl_gene_id))
```

```{r}
nrow(normalized_counts)
```

```{r}
nrow(normalized_counts) - length(which(rownames(normalized_counts) %in%
gene_mapped$ensembl_gene_id)) 
```

```{r}
normalized_counts_annot <- merge(gene_mapped,normalized_counts,
by.x = 1, by.y = 0, all.y=TRUE)
kable(normalized_counts_annot[1:10,c(1,2,3,8)],format = "html")
```

```{r}
length(ensembl_id_missing_gene <- normalized_counts_annot$ensembl_gene_id[
which(is.na(normalized_counts_annot$hgnc_symbol))])
```

## References













