---
title: "BCB420 Assignment 2"
author: "Maryam Hasanzadehkiabi"
date: "2023-03-07"
output:
  html_document:
    df_print: paged
bibliography: assignment_1.bib
csl: "american-medical-association.csl"
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Dataset selection

## Insert and running required packages

[@xie2018r]
[@allaire2023rmarkdown]
[@xie2020r]
[@Zhu2021]
[@Müller2023]
[@Wickham2022]
[@Davis2007]
[@Morgan2022]
[@zhu2008geometadb]
[@durinck2009mapping]
[@durinck2005biomart]
[@drost2017biomartr]
[@loraine2015analysis]
[@robinson2010edger]
[@Wickham2022dev]
[@Dervieux2022]
[@Xie2023kni]
[@Xie2015kni]
[@Stodden2014]
[@gu2022complex]
[@gu2016complex]
[@gu2014circlize]

```{r, message=FALSE, warning=FALSE}
if (!requireNamespace("rmarkdown", quietly = TRUE))
    +     install.packages("rmarkdown")
if (!requireNamespace("knitr", quietly = TRUE))
    +     install.packages("knitr")
if (!requireNamespace("devtools", quietly = TRUE))
    +     install.packages("devtools")
if (!requireNamespace("BiocManager", quietly = TRUE))
    +     install.packages("BiocManager")
if (!requireNamespace("edgeR", quietly = TRUE))
    BiocManager::install("edgeR")
if (!requireNamespace("biomaRt", quietly = TRUE))
    BiocManager::install("biomaRt")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")
if (!requireNamespace("GEOquery", quietly = TRUE))
    +     install.packages("GEOquery")
if (!requireNamespace("DBI", quietly = TRUE))
    +     install.packages("DBI")
if (!requireNamespace("RSQLite", quietly = TRUE))
    +     install.packages("RSQLite")
if(!requireNamespace("kableExtra", quietly=TRUE))
    install.packages("kableExtra")
if(!requireNamespace("ComplexHeatmap", quietly=TRUE))
    install.packages("ComplexHeatmap")
if(!requireNamespace("circlize", quietly=TRUE))
    install.packages("circlize")
if(!requireNamespace("gprofiler2", quietly=TRUE))
    install.packages("gprofiler2")
```

```{r message=FALSE, warning=FALSE}
library(devtools)
library(rmarkdown)
library(knitr)
library(BiocManager)
library(GEOmetadb)
library(GEOquery)
library(DBI)
library(RSQLite)
library(edgeR)
library(biomaRt)
library(kableExtra)
library(ComplexHeatmap)
library(circlize)
library(gprofiler2)
```

# Assignment 1 


## Dataset Selection Criteria

1) RNASeq data 
2) Within the last 10 years
3) Limited to human
4) Number of samples more than 6
5) Related to colon cancer

```{r message=FALSE, warning=FALSE}
sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
             " gse.submission_date,",
             " gse.supplementary_file", 
             "FROM",
             "  gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
             "  JOIN gpl ON gse_gpl.gpl=gpl.gpl",
             "WHERE",
             "  gse.submission_date > '2013-01-01' AND", 
             "  gse.title LIKE '%colon cancer%' AND", 
             "  gpl.organism LIKE '%Homo sapiens%' AND",
             "  gpl.technology LIKE '%high-throughput sequencing%' ",
             "  ORDER BY gse.submission_date DESC",sep=" ")
```

Selecting series with counts data

```{r}
if(!file.exists('GEOmetadb.sqlite')) getSQLiteFile()
con <- dbConnect(SQLite(),'GEOmetadb.sqlite')
rs <- dbGetQuery(con,sql)
counts_files <- rs$supplementary_file[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

series_of_interest <- rs$gse[grep(rs$supplementary_file,
                              pattern = "count|cnt",ignore.case = TRUE)]

shortened_filenames <- unlist(lapply(counts_files,
             FUN = function(x){x <- unlist(strsplit(x,";")) ;
              x <- x[grep(x,pattern= "count|cnt",ignore.case = TRUE)];
               tail(unlist(strsplit(x,"/")),n=1)}))
shortened_filenames[1:10]
```

The first 10 outputs of the interested series show studies focused on colorectal genomics, and hold the selection criteria. After quick reviewing, I chose GSE98922 [@schmidt2018targeting].

## Details of the Selected Series

Number of entities

```{r}
gse <- getGEO("GSE98922",GSEMatrix=FALSE)
```

Data owner contact information

```{r}
kable(data.frame(head(Meta(gse))), format = "pipe")
```

Name and number of the samples

```{r}
names(GSMList(gse))
```
Series includes 9 samples.

More information about the series

```{r}
show(getGEO("GSE98922",GSEMatrix=TRUE))
```

Samples classification

```{r}
data.frame(getGEO("GSE98922",GSEMatrix=TRUE)[[1]])[,]
```
There are 3 groups in the study, including 3 vehicle_treated, 3 AZD_treated, and 3 DBZ_treated; total 9 samples. 

## Platform details

```{r}
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))
```

**Platform accession number:**
```{r}
current_gpl_info$geo_accession
```
**Platform title:**
```{r}
current_gpl_info$title
```
**Submission date:**
```{r}
current_gpl_info$submission_date
```
**Last update date:**
```{r}
current_gpl_info$last_update_date
```
**Organisms:**
```{r}
current_gpl_info$organism
```
**Technology:**
```{r}
current_gpl_info$technology
```
**Number of GEO datasets that use this technology:*
```{r}
length(current_gpl_info$series_id)
```
**Number of GEO samples that use this technology:**
```{r}
length(current_gpl_info$sample_id)
```

# 2. Data expression

Supplementary files

```{r}
supp_files = getGEOSuppFiles('GSE98922')
file_names = rownames(supp_files)
print(file_names)
```
There are two supplementary files provided for this dataset, one includes counts data and the other one contains rlog_transformed data[@schmidt2018targeting]. Researchers studied NOTCH, MAPK, and WNT pathways signaling effect on a total 328 adenocarcinoma. Tumor cells located at the tumor edge with MAPK and WNT high activity showed epithelial-mesenchymal transition (EMT) phenotype, while tumor cells located at the tumor center showed more epithelial phenotype [@schmidt2018targeting]. Counts file is being used for further analysis.       

```{r}
genes_expression = read.delim(file_names[1],header=TRUE,
          check.names = FALSE)
```

In the following section, number of the genes and name of the variables included in the dataset are presented; since the first column includes gene names without column name, "hgnc_symbol" is assigned to the first column name.

```{r}
dim(genes_expression)
colnames(genes_expression)[1] <- "hgnc_symbol"
colnames(genes_expression)
```

Samples category

Considering the treatment agent, there are 3 groups including vehicle-treated tumors (controls), MEK inhibitor-treated (selumetinib (AZD)), and γ-secretase inhibitor-treated (dibenzazepine (DBZ)); each group contains 3 replicates.

```{r}
dataSamples <- data.frame(lapply(colnames(genes_expression)[2:10],
FUN=function(x){substr(x, 1, nchar(x)-1)}))
colnames(dataSamples) <- colnames(genes_expression)[2:10]
dataSamples <- data.frame(t(dataSamples))
dataSamples[dataSamples == "control"] <- "Vehicle"
dataSamples[dataSamples == "TreatmentA"] <- "AZD"
dataSamples[dataSamples == "TreatmentD"] <- "DBZ"
colnames(dataSamples) <- "Agent"

dataSamples_control_versus_TReatmentA <- data.frame(lapply(colnames(genes_expression)[2:7],
FUN=function(x){substr(x, 1, nchar(x)-1)}))
colnames(dataSamples_control_versus_TReatmentA) <- colnames(genes_expression)[2:7]
dataSamples_control_versus_TReatmentA <- data.frame(t(dataSamples_control_versus_TReatmentA))
dataSamples_control_versus_TReatmentA[dataSamples_control_versus_TReatmentA == "control"] <- "Vehicle"
dataSamples_control_versus_TReatmentA[dataSamples_control_versus_TReatmentA == "TreatmentA"] <- "AZD"
colnames(dataSamples_control_versus_TReatmentA) <- "Agent"

```

# 3. Genes cleaning and filtering

## Duplication
Checking potential duplication. If the logical output returns any "TRUE" value, there does exist duplication. Otherwise the set contains unique genes.

```{r}
summary(duplicated(genes_expression$hgnc_symbol))
summary(is.na(genes_expression$hgnc_symbol))
```
Since there is no "TRUE" value in the output, dataset dose not include any duplicate or missing value genes.

## Outliers
Noise filtering is performed to keep only genes with at least 1 read per million in 3 samples (i.e. smallest group of replicates).  

```{r}
cpms = cpm(genes_expression[,2:10])
rownames(cpms) <- genes_expression$hgnc_symbol
keep = rowSums(cpms > 2) >=3
genes_filtered = genes_expression[keep,]
```

## Data distribution before normalization

Boxplot

```{r}
data2plot <- log2(cpm(genes_filtered[,2:10]))
colnames(data2plot) <- colnames(genes_filtered)[2:10]
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
las = 2, cex = 0.5, cex.lab = 0.5,
cex.axis = 0.5, main = "Count Data - Boxplot", col=rainbow(10))
abline(h = median(apply(data2plot, 2, median)),
col = "yellow", lwd = 1.2, lty = "dashed")
```

Outliers with negative infinity are not drawn in plot. 

Density plot

```{r}
counts_density <- apply(log2(cpm(genes_filtered[,2:10])),
2, density)
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
xlim <- range(c(xlim, counts_density[[i]]$x));
ylim <- range(c(ylim, counts_density[[i]]$y))
}
cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
ylab="Smoothing density of log2-CPM",
main="Count Data Density Plot", cex.lab = 0.85)
for (i in 1:length(counts_density))
lines(counts_density[[i]], col=cols[i], lty=ltys[i])
legend("topright", colnames(genes_expression[2:10]),
col=cols, lty=ltys, cex=0.75,
border ="blue", text.col = "black",
merge = TRUE, bg = "gray90")
```
It seems that data follows a normal distribution pattern.

MA plot 

```{r}
plotMA(log2(genes_filtered[,c(2:4,5:7,8:10)]), ylab="M - ratio log expression",
main="Treatment vs Control")
```

#Scatter plot also does not show clustering around mean; there might be many outliers. 

# 4. Mapping

To perform mapping, hgnc_symbol is selected

```{r}
ensembl <- useMart("ensembl")
ensembl = useDataset("hsapiens_gene_ensembl",mart=ensembl)
gene_mapped <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
filters = c("hgnc_symbol"),
values = genes_filtered$hgnc_symbol,
mart = ensembl)
```


```{r}
length(which(rownames(cpms) %in%
gene_mapped$hgnc_symbol))
```
Total number of genes with measurements
```{r}
nrow(cpms)
```
Number of genes that could not be mapped
```{r}
nrow(cpms) - length(which(rownames(cpms) %in%
gene_mapped$hgnc_symbol)) 
```
Merging raw data and mapped genes; presenting first 6 rows in all cases
```{r}
cpms_annot <- merge(gene_mapped,cpms,
by.x = 2, by.y = 0, all.y=TRUE)
kable(cpms_annot[1:6,],format = "html")
```
* Another method to check ensembl ID missing genes.
```{r}
length(ensembl_id_missing_gene <- cpms_annot$ensembl_gene_id[
which(is.na(cpms_annot$ensembl_gene_id))])
```
Presenting the first 10 ensembl IDs that do not match with HUGO
```{r}
kable(cpms_annot[which(is.na(cpms_annot$ensembl_gene_id)),], format = "pipe")[1:12]
```

Old mapping

#```{r}
old_mapping <- merge(genes_expression,data.frame(ensembl_id_missing_gene), by.x = 2, by.y = 1)
kable(old_mapping[1:10,], type="pipe")
#```


# 5. Normalization

Checking potential duplication

```{r}
summary(duplicated(gene_mapped$ensembl_gene_id))
summary(duplicated(gene_mapped$hgnc_symbol))

```

Removing duplication, and creating appropriate file to perform normalization analysis

```{r}
gene_mapped_unique <- unique(gene_mapped$hgnc_symbol)
gene_mapped_unique <- data.frame(gene_mapped_unique)
rownames(gene_mapped_unique) <- gene_mapped_unique$gene_mapped_unique
gene_mapped_unique_counts = genes_expression[which(rownames(gene_mapped_unique)%in% genes_expression$hgnc_symbol),]

gene_mapped_unique_Cont_TA <- unique(gene_mapped$hgnc_symbol)
gene_mapped_unique_Cont_TA <- data.frame(gene_mapped_unique_Cont_TA)
rownames(gene_mapped_unique_Cont_TA) <- gene_mapped_unique_Cont_TA$gene_mapped_unique_Cont_TA
gene_mapped_unique_counts_Cont_TA = genes_expression[which(rownames(gene_mapped_unique_Cont_TA)%in% genes_expression$hgnc_symbol),]

```
MA Plot

```{r}
plotMA(log2(gene_mapped_unique_counts[,c(2:4,5:7,8:10)]), ylab="M - ratio log expression",main="MetS vs Control")
```

Trimmed Mean of M-values (TMM)

```{r}
filtered_data_matrix <- as.matrix(gene_mapped_unique_counts[,2:10])
rownames(filtered_data_matrix) <- gene_mapped_unique_counts$hgnc_symbol
d = DGEList(counts=filtered_data_matrix, group = dataSamples$Agent)
filtered_data_matrix_con_TA <- as.matrix(gene_mapped_unique_counts[,2:7])
rownames(filtered_data_matrix_con_TA) <- gene_mapped_unique_counts$hgnc_symbol
d_control_versus_TreatmentA <- DGEList(counts=filtered_data_matrix_con_TA, group = dataSamples_control_versus_TReatmentA$Agent)
```

Normalization factors

```{r}
d = calcNormFactors(d)
d$samples[1:9,]
d_control_versus_TreatmentA = calcNormFactors(d_control_versus_TreatmentA)
d_control_versus_TreatmentA$samples[1:6,]
#kable(d$samples[,], format = "html") %>% row_spec(c(1,8,9,10), background = "yellow")
```
The normalization factor in the highlighted samples is less than 1. It means that small number of genes are expressed more in these samples; therefore, genes with less expression have higher relative value. 

```{r}
normalized_counts <- cpm(d)
normalized_counts_vehicle_TreatmentA <- cpm(d_control_versus_TreatmentA)
```
```{r}
plotMDS(d, labels=rownames(dataSamples), col = c("darkgreen","blue")[factor(dataSamples$Agent)], main="MDS Plot")

plotMDS(d_control_versus_TreatmentA, labels=rownames(dataSamples_control_versus_TReatmentA), col = c("darkgreen","blue")[factor(dataSamples_control_versus_TReatmentA$Agent)], main="MDS Plot")

```
Samples are not cluster with each other, there are unpaired samples that show different expression. It seems that N3, M1, M2, M3 and M6 samples could be less different. 

```{r}
model_design <- model.matrix(~dataSamples$Agent)
```
```{r}
d <- estimateDisp(d, model_design)
```
```{r}
plotBCV(d,col.tagwise = "black",col.common = "red",col.trend = "blue", main="Biological Coeffient of Variation")
```



# Assignment 2

## 1.2. Preparing dataset for further analysis

Since my selected dataset that was used in my assignment 1 was not appropriate, I should redo my assignment 1 with the new selected dataset. My selection criteria includes:
1) dataset with RNASeq data 
2) Within the last 10 years
3) Limited to human
4) Number of samples more than 6
5) Related to colon cancer
Therefore, I did part of assignment 1 again to get normalized count data to be able to complete my assignment 2. The new selected dataset is GSE98922, considering the treatment agent the patients are divided to vehicle-treated tumors (controls), MEK inhibitor-treated (selumetinib (AZD)), and γ-secretase inhibitor-treated (dibenzazepine (DBZ)); each group contains 3 replicates [@schmidt2018targeting]. By checking MDS plot, vehicle group and DBZ group are clustered very closely, therefore, further anaylyses were performed over vehicle and AZD groups. 

```{r}
normalized_count_annot_Cont_TA <- merge(gene_mapped,
normalized_counts_vehicle_TreatmentA, by.x = 2, by.y = 0, all.y = TRUE) 

summary(duplicated(normalized_count_annot_Cont_TA$ensembl_gene_id)) #checking duplicated gene_id
normalized_count_annot_Cont_TA_unique_gene_id <- subset(normalized_count_annot_Cont_TA, !duplicated(normalized_count_annot_Cont_TA$ensembl_gene_id)) #unique gene_id
summary(is.na(normalized_count_annot_Cont_TA_unique_gene_id$ensembl_gene_id)) #there is one missing value
normalized_count_annot_Cont_TA_unique_gene_id <- subset(normalized_count_annot_Cont_TA_unique_gene_id, !is.na(normalized_count_annot_Cont_TA_unique_gene_id$ensembl_gene_id)) #missing value removed

summary(duplicated(normalized_count_annot_Cont_TA$hgnc_symbol)) #checking duplicated hgnc_symbol
normalized_count_annot_Cont_TA_unique_hgnc_symbol <- subset(normalized_count_annot_Cont_TA, !duplicated(normalized_count_annot_Cont_TA$hgnc_symbol)) #unique hgnc_symbol
summary(is.na(normalized_count_annot_Cont_TA_unique_hgnc_symbol$hgnc_symbol)) #there is no missing value

```

# 2. Differential Gene Expresion

## 2.1. Heatmap 

```{r}
heatmap_matrix_gene_id <- normalized_count_annot_Cont_TA_unique_gene_id[,3:8] # creating a numerical matrix over unique gene_id
rownames(heatmap_matrix_gene_id) <- normalized_count_annot_Cont_TA_unique_gene_id$ensembl_gene_id
colnames(heatmap_matrix_gene_id) <- colnames(normalized_count_annot_Cont_TA_unique_gene_id[,3:8])
```

## 2.2. Translating gene numbers to a color scale

```{r}
heatmap_matrix_gene_id <- t(scale(t(heatmap_matrix_gene_id)))
if(min(heatmap_matrix_gene_id) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_gene_id)),
c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix_gene_id), 0,
max(heatmap_matrix_gene_id)), c("blue", "white", "red"))
}
current_heatmap_gene_id <- Heatmap(as.matrix(heatmap_matrix_gene_id),
show_row_dend = TRUE,show_column_dend = TRUE,
col=heatmap_col,show_column_names = TRUE,
show_row_names = FALSE,show_heatmap_legend = TRUE)
current_heatmap_gene_id
```

## 2.3. Compairing the expression of interested genes

```{r}
controls_samples_gene_id <- grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^c")
TreatmentA_samples_gene_id <- grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^T")
interested_gene <- which(normalized_count_annot_Cont_TA_unique_gene_id$hgnc_symbol== "FN1")
FN1_control_samples <-
t(normalized_count_annot_Cont_TA_unique_gene_id
[interested_gene,
controls_samples_gene_id])
colnames(FN1_control_samples) <-
c("control_samples")
FN1_control_samples

FN1_Treatment_samples <-
t(normalized_count_annot_Cont_TA_unique_gene_id
[interested_gene,
TreatmentA_samples_gene_id])
colnames(FN1_Treatment_samples) <-
c("Treatment_samples")
FN1_Treatment_samples

t.test(x=t(FN1_control_samples),y=t(FN1_Treatment_samples))

interested_gene_2 <- which(normalized_count_annot_Cont_TA_unique_gene_id$hgnc_symbol== "CDH1")

CDH1_control_samples <-
t(normalized_count_annot_Cont_TA_unique_gene_id
[interested_gene_2,
controls_samples_gene_id])
colnames(CDH1_control_samples) <-
c("control_samples")
CDH1_control_samples

CDH1_Treatment_samples <-
t(normalized_count_annot_Cont_TA_unique_gene_id
[interested_gene_2,
TreatmentA_samples_gene_id])
colnames(CDH1_Treatment_samples) <-
c("Treatment_samples")
CDH1_Treatment_samples

t.test(x=t(CDH1_control_samples),y=t(CDH1_Treatment_samples))

```
The primary investigators reported that CDH1 and FN1 genes are expressed differently in two treated groups. That is why I selected these 2 genes to investigate their expression over control and AZD groups. t-test analysis revealed that FN1 expression is not significantly different (t_test=0.61725, p-value=0.5747), however CDH1 expression was detected to be expressed significantly different between 2 groups (t-test=t = -9.439, p-value=0.004122).         

# 3.Data matrix

First I checked how the samples are clustering.

```{r}
limma::plotMDS(heatmap_matrix_gene_id, col = c("darkgreen","blue")[factor(dataSamples$Agent)], main="MDS Plot")
```
MDS plot shows that 2 groups are clearly clustered separately.
In the following section, expression matrix is created.

```{r}
gene_id_expressionMatrix <- as.matrix(normalized_count_annot_Cont_TA_unique_gene_id[,3:8])
rownames(gene_id_expressionMatrix) <-
normalized_count_annot_Cont_TA_unique_gene_id$ensembl_gene_id
colnames(gene_id_expressionMatrix) <-
colnames(normalized_count_annot_Cont_TA_unique_gene_id)[3:8]
```

A linear model is created for performing further analyses.

```{r}
gene_id_modelDesign <- model.matrix(~ dataSamples_control_versus_TReatmentA$Agent)
gene_id_minimalSet <- ExpressionSet(assayData=gene_id_expressionMatrix)
gene_id_fit_simpleModel <- lmFit(gene_id_minimalSet, gene_id_modelDesign)
```

Data is fitted to the simple model.

```{r}
gene_id_minimalSet <- ExpressionSet(assayData=gene_id_expressionMatrix)
gene_id_fit <- lmFit(gene_id_minimalSet, gene_id_modelDesign)
```

Computing differential expression using emprical Bayes

```{r}
gene_id_fit2 <- eBayes(gene_id_fit,trend=TRUE)
gene_id_topfit <- topTable(gene_id_fit2,
coef=ncol(gene_id_modelDesign),
adjust.method = "BH",
number = nrow(gene_id_expressionMatrix))
gene_id_output_hits <- merge(normalized_count_annot_Cont_TA_unique_gene_id[,1:2],
gene_id_topfit,
by.y=0,by.x=2,
all.y=TRUE)
gene_id_output_hits <- gene_id_output_hits[order(gene_id_output_hits$P.Value),]
```

## 3.1. Number of genes with p-value less than 0.05

```{r}
numberOfGenes_pvalue_less_0.05 <- length(which(gene_id_output_hits$P.Value < 0.05))
numberOfGenes_Adjusted_pvalue_less_0.05 <- length(which(gene_id_output_hits$adj.P.Val < 0.05))
kable(gene_id_output_hits[1:10,],type="html",row.names = FALSE)
```

# 4. Thresholded over-representation analysis

## 4.1. Up regulated and down regulated genes

Since there are thousands genes in the dataset, to get strong ones, the threshold is set to 1*10^-6.

```{r}
cpms_cont_TA = cpm(genes_expression[,c(2:7)])
rownames(cpms_cont_TA) <- genes_expression$hgnc_symbol
keep = rowSums(cpms_cont_TA > 1) >=3
genes_filtered_cont_TA = genes_expression[keep,c(2:7)]

filtered_data_matrix <- as.matrix(genes_filtered[,2:7])
rownames(filtered_data_matrix) <- genes_filtered$hgnc_symbol

d_control_versus_TreatmentA <- DGEList(counts=filtered_data_matrix_con_TA, group = dataSamples_control_versus_TReatmentA$Agent)

d_control_versus_TreatmentA <- estimateDisp(d_control_versus_TreatmentA, gene_id_modelDesign)
d_control_versus_TreatmentA <- estimateCommonDisp(d_control_versus_TreatmentA, gene_id_modelDesign) # estimating dispersion
d_control_versus_TreatmentA <- calcNormFactors(d_control_versus_TreatmentA) # normalization factors calculation

fit_edgeR <- glmQLFit(d_control_versus_TreatmentA, gene_id_modelDesign)
qlf.control_versus_TreatmentA <- glmQLFTest(fit_edgeR)
qlf_output_hits <- topTags(qlf.control_versus_TreatmentA, sort.by = "PValue", n=nrow(filtered_data_matrix))

length(which(qlf_output_hits$table$PValue < 0.000001))# number of genes with p-value < 0.000001
length(which(qlf_output_hits$table$FDR < 0.000001)) # number of genes with adjusted p_value (FDR) < 0.000001

length(which(qlf_output_hits$table$PValue < 0.000001
& qlf_output_hits$table$logFC > 0)) # number of up regulated genes
length(which(qlf_output_hits$table$PValue < 0.000001
& qlf_output_hits$table$logFC < 0)) # number of down regulated genes
```

## 4.2. Thresholded lists

```{r}
qlf_output_hits_withgn <- merge(cpms_annot[,1:2],qlf_output_hits, by.x=1, by.y=0, all.y = TRUE)
qlf_output_hits_withgn[,"rank"] <- -log(qlf_output_hits_withgn$PValue, base=10)*sign(qlf_output_hits_withgn$logFC)
qlf_output_hits_withgn <- qlf_output_hits_withgn[order(qlf_output_hits_withgn$rank),]
upregulated_genes <- qlf_output_hits_withgn$hgnc_symbol[
which(qlf_output_hits_withgn$PValue < 0.000001
& qlf_output_hits_withgn$logFC > 0)]
downregulated_genes <- qlf_output_hits_withgn$hgnc_symbol[
which(qlf_output_hits_withgn$PValue < 0.000001
& qlf_output_hits_withgn$logFC < 0)]
write.table(x=upregulated_genes,
file=file.path("CRC_upregulated_genes.txt"),sep="\t",
row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=downregulated_genes,
file=file.path("CRC_downregulated_genes.txt"),sep="\t",
row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=data.frame(genename= qlf_output_hits_withgn$hgnc_symbol,F_stat= qlf_output_hits_withgn$rank),
file=file.path("CRC_ranked_genelist.txt"),sep="\t",
row.names = FALSE,col.names = FALSE,quote = FALSE)
```



# Differential Gene Expression Questions

**1. Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?**

Initially, the threshold was kept as the convenient value 0.05; there are 2,973 genes with p-value less than 0.05, in the dataset.

**2. Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?**

Benjamni - hochberg (BH) method was used to handle multiple testing. False discovery rate (FDR) can be controlled when multiple hypotheses are testing simultaneously by adjusting p-value for each hypothesis [@benjamini1995controlling]; it accounts for high chance of false positive. There are 2,682 genes with adjusted p-value (FDR) less than 0.05.

**3. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.**

```{r}
Control <- normalized_count_annot_Cont_TA_unique_gene_id[,grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^c")]
Treatment_A <- normalized_count_annot_Cont_TA_unique_gene_id[,grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^T")]
avg <- data.frame(hgnc_symbol = normalized_count_annot_Cont_TA_unique_gene_id[,1],
                  Control = rowMeans(Control),
                  Treatment_A = rowMeans(Treatment_A))

grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^c")
TreatmentA_samples_gene_id <- grep(colnames(normalized_count_annot_Cont_TA_unique_gene_id), pattern = "^T")

status <- rep(0, nrow(qlf_output_hits_withgn))
status[qlf_output_hits_withgn$logFC < 0 & qlf_output_hits_withgn$PValue < 0.05] <- -1
status[qlf_output_hits_withgn$logFC > 0 & qlf_output_hits_withgn$PValue < 0.05] <- 1

limma::plotMD(log2(qlf_output_hits_withgn[,c(3,6)]),
              status=status,
              values=c(-1,1),
              hl.col=c("blue","red"),
              main="MA Plot")
```

**4. Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not. Make sure all your figures have proper heading and labels. Every figure included in the report should have a detailed figure legend.**

```{r}
tophits_gene_id <- gene_id_output_hits$ensembl_gene_id[gene_id_output_hits$P.Value<0.01]
tophits_heatmap_matrix_gene_id <- t(
scale(t(heatmap_matrix_gene_id[which(rownames(heatmap_matrix_gene_id)
%in% tophits_gene_id),])))
if(min(heatmap_matrix_gene_id) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_gene_id)),
c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix_gene_id), 0,
max(heatmap_matrix_gene_id)), c("blue", "white", "red"))
}
current_tophits_heatmap_gene_id <- Heatmap(as.matrix(tophits_heatmap_matrix_gene_id),
cluster_rows = TRUE, show_row_dend = TRUE,
cluster_columns = FALSE, show_column_dend = FALSE,
col=heatmap_col,show_column_names = TRUE,
show_row_names = FALSE,show_heatmap_legend = TRUE)
current_tophits_heatmap_gene_id
```
Heatmap clearly shows different expression of genes between 2 groups.

# Thresholded over-representation analysis questions

**1. Which method did you choose and why?**

g:Profiler website is used due to its easy and user friendly annotating tool.

**2. What annotation data did you use and why? What version of the annotation are you using?**

GO:BP, Reactome, and WikiPathways. versione108_eg55_p17_0254fbf. 

**3. How many genesets were returned with what thresholds?**

For up_regulated genes: GO:BP # 8, Reactome # 3, threshold: 0.000001
For down_regulated genes: it does not return any dataset with threshold: 0.000001

**4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated
and down regulated differentially expressed genes separately)?**

Since there are more than 10,000 genes in whole list, g:profiler website could not revealed results; therefore I could not compare whole list with up and down regulated ones.




## References