---
title: "Assignment 3"
author: "Maryam Hasanzadehkiabi"
output: html_notebook
bibliography: bibliography.bib
csl: american-medical-association.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Insert and running required packages

[@xie2018r]
[@allaire2023rmarkdown]
[@xie2020r]
[@Zhu2021]
[@Müller2023]
[@Wickham2022]
[@Davis2007]
[@Morgan2022]
[@zhu2008geometadb]
[@durinck2009mapping]
[@durinck2005biomart]
[@drost2017biomartr]
[@loraine2015analysis]
[@robinson2010edger]
[@Wickham2022dev]
[@Dervieux2022]
[@Xie2023kni]
[@Xie2015kni]
[@Stodden2014]

```{r, message=FALSE, warning=FALSE}
if (!requireNamespace("rmarkdown", quietly = TRUE))
    +     install.packages("rmarkdown")
if (!requireNamespace("knitr", quietly = TRUE))
    +     install.packages("knitr")
if (!requireNamespace("devtools", quietly = TRUE))
    +     install.packages("devtools")
if (!requireNamespace("BiocManager", quietly = TRUE))
    +     install.packages("BiocManager")
if (!requireNamespace("edgeR", quietly = TRUE))
    BiocManager::install("edgeR")
if (!requireNamespace("biomaRt", quietly = TRUE))
    BiocManager::install("biomaRt")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")
if (!requireNamespace("GEOquery", quietly = TRUE))
    +     install.packages("GEOquery")
if (!requireNamespace("DBI", quietly = TRUE))
    +     install.packages("DBI")
if (!requireNamespace("RSQLite", quietly = TRUE))
    +     install.packages("RSQLite")
if(!requireNamespace("kableExtra", quietly=TRUE))
    install.packages("kableExtra")
if (!requireNamespace("RCurl", quietly = TRUE))
    +     install.packages("RCurl")
if (!requireNamespace("RCy3", quietly = TRUE))
    +     install.packages("RCy3")
if (!requireNamespace("ggplot2", quietly = TRUE))
    +     install.packages("ggplot2")


```

```{r message=FALSE, warning=FALSE}
library(devtools)
library(rmarkdown)
library(knitr)
library(BiocManager)
library(GEOmetadb)
library(GEOquery)
library(DBI)
library(RSQLite)
library(edgeR)
library(biomaRt)
library(kableExtra)
library(RCurl)
library(RCy3)
library(ggplot2)
```

# Non-thresholded Gene set Enrichment Analysis

**1. What method did you use? What genesets did you use? Make sure to specify versions and cite your methods.**

1) “GSEA v4.3.2” was used to perform non-thresholded  gene set enrichment analysis [@subramanian2005gene], [@mootha2003pgc].

2) “Human_GOBP_AllPathways_no_GO_iea_April_02_2023_symbol.gmt” was downloaded manually from the Bader lab, and used as genesets. 

3) CRC_ranked_genelist.rank from assignment 2, was used as the ranked list. Other parameters include: number of permutations: 1,000; No_Collapse; Max size: 200; and Min size: 15.

**2. Summarize your enrichment results.**

The GSEA report is presented in Figure 1. Filtering (min: 15, max:200) results in 3,945 gene sets that are included in the analysis.

**Figure 1. GSEA Report Summary**
```{r echo=FALSE, out.width = "50%", fig.align = "center"}
knitr::include_graphics("/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/gsea_report.png")
```

In the following sections, the top 5 pathways in each group (controls versus TreatmentA) are summarized in Table 1 and Table 2.

```{r}
controls_g <- read.delim("/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/control_vs_treatmentA.GseaPreranked.1680983057326/gsea_report_for_na_pos_1680983057326.tsv", header = TRUE)

controls_g <- controls_g[, !(names(controls_g) %in% c("GS.br..follow.link.to.MSigDB", "GS.DETAILS", "X"))]
kable(head(controls_g, n = 5), type="html", caption = "Table 1. Top 5 positive enrichment GSEA results for controls", row.names = FALSE) %>%
  kable_styling()
```

```{r}
treatmentA_g <- read.delim("/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/control_vs_treatmentA.GseaPreranked.1680983057326/gsea_report_for_na_neg_1680983057326.tsv", header = TRUE)

treatmentA_g <- treatmentA_g[, !(names(controls_g) %in% c("GS.br..follow.link.to.MSigDB", "GS.DETAILS", "X"))]
kable(head(treatmentA_g, n = 5), type="html", caption = "Table 2. Top 5 Negative enrichment GSEA results for treatmentA", row.names = FALSE) %>%
  kable_styling()

```

**3. How do these results compare to the results from the thresholded analysis in Assignment #2. Compare qualitatively. Is this a straight forward comparison? Why or why not?**

In the assignment 2, g:Profiler reports separately for up-regulated and down-regulated gene list, how genes are differentially expressed. While in GSEA, a ranked gene list including both up-regulated and down-regulated genes is used to get genes expression.

g:Profiler detected 8 pathways annotated by GO:BP, 3 pathways annotated by Reactome, and non annotated by WikiPathways in up-regulated genes, the top one of each is listed below:  
protein folding (GO:00064574)
PERK regulates gene expression (REAC:R-HSA-381042)
With respect to the down-regulated genes, no results was returned by g:Profiler.

The top pathway detected by GSEA is
RIBONUCLEOPROTEIN COMPLEX BIOGENESIS (GO:0022613) in control group, and 
DEFENSINS%REACTOME DATABASE (ID RELEASE 1461973) in TreatmentA group.

Since the input and output files are different in each method (g:Profiler vs GSEA), they could not be compared easily.

# Visualize your Gene set Enrichment Analysis in Cytoscape

**1. Create an enrichment map - how many nodes and how many edges in the resulting map? What thresholds were used to create this map? Make sure to record all thresholds. Include a screenshot of your network prior to manual layout.**

Cytoscope v3.9.1 was used to create enrichment map [@shannon2003cytoscape], [@otasek2019cytoscape], [@reimand2019pathway].

Thresholds that used include FDR q-value: 0.01, Filter genes by expressions was selected, connectivity slider for the number of edges was hold in the center, similarity metric: Jaccard (50%), Overlap (50%)combined (0.375)
In the resulted map, there are 111 nodes and 367 edges.

A screenshot is displayed below.

```{r}
EM_criteria <- file.path(getwd(),"EM_criteria.png")
response <- exportImage(EM_criteria, type = "png")

htmltools::img(src = knitr::image_uri(EM_criteria), 
               alt = 'Enrichment Map Criteria', 
               style = 'margin:0px auto;display:block')
```


```{r}
gmt_url = "http://download.baderlab.org/EM_Genesets/current_release/Human/symbol/"
```
```{r}
filenames = getURL(gmt_url)
tc = textConnection(filenames)
contents = readLines(tc)
close(tc)
```

```{r}
rx = gregexpr("(?<=<a href=\")(.*.GOBP_AllPathways_no_GO_iea.*.)(.gmt)(?=\">)", contents,
    perl = TRUE)
gmt_file = unlist(regmatches(contents, rx))
dest_gmt_file <- file.path("/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/data_dir", gmt_file)
download.file(paste(gmt_url, gmt_file, sep = ""), destfile = dest_gmt_file)
```


```{r}
gsea_jar <- "/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/GSEA_4.3.2/gsea-cli.sh"
java_version <- 12
run_gsea = "/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/GSEA_4.3.2/gsea.sh"

working_directory = setwd("/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/control_vs_treatmentA.GseaPreranked.1680983057326")
gsea_directory = paste(working_directory,"control_vs_treatmentA.GseaPreranked.1680983057326", sep = "/")

analysis_name <- "/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/control_vs_treatmentA.GseaPreranked.1680983057326/control_vs_treatmentA.GseaPreranked.1680983057326.rpt"

rank_file <- "/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/result_data/control_vs_treatmentA.GseaPreranked.1680983057326/ranked_gene_list_na_pos_versus_na_neg_1680983057326.tsv"

espression_file <- "/Users/maryamhasanzadehkiabi/Documents/BCB420/Maryam_Hasanzadehkiabi/Assignment 3/project_data/CRC_ranked_genelist.rnk" 

command <- paste("java  -Xmx1G -cp",gsea_jar,  "xtools.gsea.GseaPreranked -gmx", dest_gmt_file, "-rnk" ,file.path(working_directory,rank_file), "-collapse false -nperm 1000 -permute gene_set -scoring_scheme weighted -rpt_label ",analysis_name,"  -num 100 -plot_top_x 20 -rnd_seed 12345  -set_max 200 -set_min 15 -zip_report false -out" ,working_directory, "-gui false > gsea_output.txt",sep=" ")

gsea_output_dir <- gsea_directory

cytoscapePing()
cytoscapeVersionInfo()

pvalue_gsea_threshold <- "0.05"
qvalue_gsea_threshold <- "0.01"
similarity_threshold <- "0.375"
similarity_metric = "COMBINED"
cur_model_name <- analysis_name
gsea_results_path <- file.path(gsea_output_dir,"edb")
gsea_results_filename <- file.path(gsea_results_path,"results.edb")

gmt_gsea_file <- file.path(getwd(),dest_gmt_file)
gsea_ranks_file <- file.path(gsea_results_path,list.files(gsea_results_path,pattern=".rnk"))


current_network_name <- paste(cur_model_name,pvalue_gsea_threshold,qvalue_gsea_threshold,sep="_")
em_command = paste('enrichmentmap build analysisType="gsea" gmtFile=',gmt_gsea_file,
                   'pvalue=',pvalue_gsea_threshold, 'qvalue=',qvalue_gsea_threshold,
                   'similaritycutoff=',similarity_threshold,
                   'coefficients=',similarity_metric,'ranksDataset1=', 
                   gsea_ranks_file,'enrichmentsDataset1=',gsea_results_filename, 
                   'filterByExpressions=false',
                   'expressionDataset1=',file.path(getwd(),working_directory,espression_file),
                   sep=" ")


#response <- commandsGET(em_command)
current_network_suid <- 0


current_names <- getNetworkList()
output_network_file <- file.path(getwd(),"initial_screenshot_network.png")
response <- exportImage(output_network_file, type = "png")

htmltools::img(src = knitr::image_uri(output_network_file), 
               alt = 'Initial Enrichment Map', 
               style = 'margin:0px auto;display:block')
```

**2. Annotate your network - what parameters did you use to annotate the network. If you are using the default parameters make sure to list them as well.**

An AutoAnnotated output was created and is displayed below (screen shot).

```{r}
AutoAnnotated <- file.path(getwd(),"AutoAnnotated.png")
response <- exportImage(output_network_file, type = "png")

htmltools::img(src = knitr::image_uri(AutoAnnotated), 
               alt = 'Initial Enrichment Map', 
               style = 'margin:0px auto;display:block')
```

**3. Make a publica"on ready figure - include this figure with proper legends in your notebook.**


Another annotation using parameters as Cluster algorithm: MCL Cluster, Edge weight column: similarity_coefficient, singleton clusters, prevent cluster overlap, Lable column: GS_DESCR, Lable algorithm: WordCloud (Adjacent Words), max words: 3, min word: 1, Adjacent word: 8, was performed and the screen shot is shown below

```{r}
Annotated <- file.path(getwd(),"Annotated.png")
response <- exportImage(output_network_file, type = "png")

htmltools::img(src = knitr::image_uri(Annotated), 
               alt = 'Initial Enrichment Map', 
               style = 'margin:0px auto;display:block')
```

**4. Collapse your network to a theme network. What are the major themes present in this analysis? Do they fit with the model? Are there any novel pathways or themes?**


By using filter in control panel, a sub-network was created. The new sub-network includes 2 clusters rRNA processing cytosol that contains 22 genes, and RNA 3' end that contains 2 genes. It is displayed below

```{r}
subnet_annotation <- file.path(getwd(),"subnet_annotation.png")
response <- exportImage(output_network_file, type = "png")

htmltools::img(src = knitr::image_uri(subnet_annotation), 
               alt = 'Initial Enrichment Map', 
               style = 'margin:0px auto;display:block')
```

# Interpreta"on and detailed view of results

**1. Do the enrichment results support conclusions or mechanism discussed in the original paper? How do these results differ from the results you got from Assignment #2 thresholded methods**

I could not get even close results to what the primary investigators indicated in the original article.


**2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your result?**


**1. Add a post analysis to your main network using specific transcrip"on factors, microRNAs or drugs. Include the reason why you chose the specific miRs, TFs or drugs (i.e publications indicating that they might be related to your
model). What does this post analysis show?**

Since the main study was conducted to determine the effect of some drugs on different groups of colon cancer [@schmidt2018targeting], a post analysis using drugs was performed over the network. Human_DrugBank_all_symbol was downloaded and used as signature gene sets. Some drugs such as COPPER, CHOLIC ACID, GELDANAMYCIN, and ARTENIMOL were found as significantly interacted with different genes and pathways in the study (screenshot is provided below). There are few studies that provide supports for these substances in the treatment of cancers [@ni2019prediction], [@barka2016taxonomy], [@sung2023natural], [@hammood2021assess].

```{r}
drugbank <- file.path(getwd(),"drugbank.png")
response <- exportImage(output_network_file, type = "png")

htmltools::img(src = knitr::image_uri(drugbank), 
               alt = 'Post Analysis', 
               style = 'margin:0px auto;display:block')
```



This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
